File: python-examples/computer-use/lib/openai/agent/agent.py
import json
from collections.abc import Callable

from lib.openai.agent.utils import create_response, sanitize_message
from lib.openai.computers import Computer


class Agent:
    """
    An async agent class that can be used to interact with a computer using OpenAI's API.
    """

    def __init__(
        self,
        model="gpt-5.4",
        api_key: str = None,
        base_url: str = None,
        computer: Computer = None,
        tools: list[dict] | None = None,
        acknowledge_safety_check_callback: Callable = lambda message: True,
    ):
        self.model = model
        self.api_key = api_key
        self.base_url = base_url
        self.computer = computer
        self.tools = list(tools or [])
        self.print_steps = True
        self.debug = False
        self.show_images = False
        self.acknowledge_safety_check_callback = acknowledge_safety_check_callback

        if computer:
            self.tools.append({"type": "computer"})

    def debug_print(self, *args):
        if self.debug:
            print(json.dumps(args, indent=4))

    async def handle_item(self, item):
        """Handle each item; may cause a computer action + screenshot."""
        if item["type"] == "message":
            if self.print_steps:
                text_blocks = [
                    block["text"]
                    for block in item.get("content", [])
                    if isinstance(block, dict) and "text" in block
                ]
                if text_blocks:
                    print("\n".join(text_blocks))

        if item["type"] == "function_call":
            name, args = item["name"], json.loads(item["arguments"])
            if self.print_steps:
                print(f"{name}({args})")

            if hasattr(self.computer, name):  # if function exists on computer, call it
                method = getattr(self.computer, name)
                await method(**args)
            return [
                {
                    "type": "function_call_output",
                    "call_id": item["call_id"],
                    "output": "success",
                }
            ]

        if item["type"] == "computer_call":
            actions = item.get("actions") or (
                [item["action"]] if "action" in item else []
            )
            for action in actions:
                action_type = action["type"]
                action_args = {k: v for k, v in action.items() if k != "type"}
                if self.print_steps:
                    print(f"{action_type}({action_args})")

                method = getattr(self.computer, action_type)
                await method(**action_args)

            screenshot_base64 = await self.computer.screenshot()

            # if user doesn't ack all safety checks exit with error
            pending_checks = item.get("pending_safety_checks", [])
            for check in pending_checks:
                message = check["message"]
                if not self.acknowledge_safety_check_callback(message):
                    raise ValueError(
                        f"Safety check failed: {message}. Cannot continue with unacknowledged safety checks."
                    )

            call_output = {
                "type": "computer_call_output",
                "call_id": item["call_id"],
                "output": {
                    "type": "computer_screenshot",
                    "image_url": f"data:image/png;base64,{screenshot_base64}",
                    "detail": "original",
                },
            }

            return [call_output]
        return []

    async def run_full_turn(
        self, input_items, print_steps=True, debug=False, show_images=False
    ):
        self.print_steps = print_steps
        self.debug = debug
        self.show_images = show_images
        new_items = []
        current_input = input_items
        previous_response_id = None

        # keep looping until we get a final response
        while new_items[-1].get("role") != "assistant" if new_items else True:
            self.debug_print(
                [sanitize_message(msg) for msg in current_input]
                + (
                    [{"previous_response_id": previous_response_id}]
                    if previous_response_id
                    else []
                )
            )

            response = create_response(
                api_key=self.api_key,
                base_url=self.base_url,
                model=self.model,
                input=current_input,
                previous_response_id=previous_response_id,
                tools=self.tools,
            )
            self.debug_print(response)
            previous_response_id = response.get("id")

            if "output" not in response and self.debug:
                print(response)
                raise ValueError("No output from model")
            else:
                current_input = []
                new_items += response["output"]
                for item in response["output"]:
                    handled_items = await self.handle_item(item)
                    new_items += handled_items
                    current_input += handled_items

        return new_items

File: typescript-examples/computer-use/lib/openai/agent.ts
import OpenAI from 'openai';
import type {
  ResponseItem,
  ResponseInputItem,
  ResponseOutputMessage,
  ResponseOutputText,
  ResponseFunctionToolCallItem,
  ResponseFunctionToolCallOutputItem,
  ResponseComputerToolCall,
  ResponseComputerToolCallOutputItem,
  Tool,
} from 'openai/resources/responses/responses.js';

import * as utils from './utils';
import type { PlaywrightComputer } from './computer';

type ComputerAction = {
  type: string;
  [key: string]: unknown;
};

export class Agent {
  private model: string;
  private computer: PlaywrightComputer;
  private tools: Tool[];
  private print_steps = true;
  private debug = false;
  private show_images = false;
  private openai: OpenAI;
  private ackCb: (msg: string) => boolean;

  constructor(opts: {
    model?: string;
    computer: PlaywrightComputer;
    tools?: Tool[];
    acknowledge_safety_check_callback?: (msg: string) => boolean;
    apiKey: string;
    baseURL?: string;
  }) {
    this.model = opts.model ?? 'gpt-5.4';
    this.computer = opts.computer;
    this.openai = new OpenAI({ apiKey: opts.apiKey, baseURL: opts.baseURL });
    this.tools = [...(opts.tools ?? [])];
    this.ackCb = opts.acknowledge_safety_check_callback ?? ((): boolean => true);

    this.tools.push({ type: 'computer' } as unknown as Tool);
  }

  private debugPrint(...args: unknown[]): void {
    if (this.debug) {
      console.warn('--- debug:agent:debugPrint');
      try {
        console.dir(
          args.map((msg) => utils.sanitizeMessage(msg as ResponseItem)),
          { depth: null },
        );
      } catch {
        console.dir(args, { depth: null });
      }
    }
  }

  private async handleItem(item: ResponseItem): Promise<ResponseItem[]> {
    if (item.type === 'message' && this.print_steps) {
      const msg = item as ResponseOutputMessage;
      const c = msg.content;
      const textBlocks = Array.isArray(c)
        ? c
            .filter(
              (block): block is ResponseOutputText =>
                block.type === 'output_text' && typeof block.text === 'string',
            )
            .map((block) => block.text)
        : [];
      if (textBlocks.length > 0) console.log(textBlocks.join('\n'));
    }

    if (item.type === 'function_call') {
      const fc = item as ResponseFunctionToolCallItem;
      const argsObj = JSON.parse(fc.arguments) as Record<string, unknown>;
      if (this.print_steps) console.log(`${fc.name}(${JSON.stringify(argsObj)})`);
      
      const fn = (this.computer as unknown as Record<string, unknown>)[fc.name];
      if (typeof fn === 'function') {
        await (fn as (args: Record<string, unknown>) => unknown)(argsObj);
      }
      
      return [
        {
          type: 'function_call_output',
          call_id: fc.call_id,
          output: 'success',
        } as unknown as ResponseFunctionToolCallOutputItem,
      ];
    }

    if (item.type === 'computer_call') {
      const cc = item as ResponseComputerToolCall;
      const actions = ((cc as unknown as { actions?: ComputerAction[] }).actions ??
        [cc.action]) as ComputerAction[];
      for (const action of actions) {
        const { type: actionType, ...actionArgs } = action;
        if (this.print_steps) console.log(`${actionType}(${JSON.stringify(actionArgs)})`);

        const fn = (this.computer as unknown as Record<string, unknown>)[actionType];
        if (typeof fn === 'function') {
          await (fn as (args: Record<string, unknown>) => unknown)(actionArgs);
        }
      }

      const screenshot = await this.computer.screenshot();
      const pending = cc.pending_safety_checks ?? [];
      for (const { message } of pending) {
        if (message && !this.ackCb(message)) {
          throw new Error(`Safety check failed: ${message}`);
        }
      }

      const out: Omit<ResponseComputerToolCallOutputItem, 'id'> = {
        type: 'computer_call_output',
        call_id: cc.call_id,
        output: {
          type: 'computer_screenshot',
          image_url: `data:image/png;base64,${screenshot}`,
        },
      };
      return [out as ResponseItem];
    }

    return [];
  }

  async runFullTurn(opts: {
    messages: ResponseInputItem[];
    print_steps?: boolean;
    debug?: boolean;
    show_images?: boolean;
  }): Promise<ResponseItem[]> {
    this.print_steps = opts.print_steps ?? true;
    this.debug = opts.debug ?? false;
    this.show_images = opts.show_images ?? false;
    const newItems: ResponseItem[] = [];
    let currentInput: ResponseInputItem[] = [...opts.messages];
    let previousResponseId: string | null = null;

    while (
      newItems.length === 0 ||
      (newItems[newItems.length - 1] as ResponseItem & { role?: string }).role !== 'assistant'
    ) {
      this.debugPrint(
        ...currentInput,
        ...(previousResponseId ? [{ previous_response_id: previousResponseId }] : []),
      );
      const response = await utils.createResponse(this.openai, {
        model: this.model,
        input: currentInput,
        previous_response_id: previousResponseId,
        tools: this.tools as OpenAI.Responses.Tool[],
      } as OpenAI.Responses.ResponseCreateParams);
      if (!response.output) throw new Error('No output from model');
      previousResponseId =
        typeof (response as { id?: unknown }).id === 'string'
          ? ((response as { id: string }).id)
          : null;
      currentInput = [];
      for (const msg of response.output as ResponseItem[]) {
        const handledItems = await this.handleItem(msg);
        newItems.push(msg, ...handledItems);
        currentInput.push(...(handledItems as ResponseInputItem[]));
      }
    }

    // Return sanitized messages if show_images is false
    return !this.show_images
      ? newItems.map((msg) => utils.sanitizeMessage(msg) as ResponseItem)
      : newItems;
  }
}
